File: tensorflow_lite_support/examples/task/vision/desktop/image_classifier_demo.c
/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

// Example usage:
// bazel run -c opt \
//  tensorflow_lite_support/examples/task/vision/desktop:image_classifier_demo \
//  -- \
//  --model_path=/path/to/model.tflite \
//  --image_path=/path/to/image.jpg

#include <stdio.h>
#include <stdint.h>

#include "tensorflow_lite_support/cc/task/vision/image_classifier_c_api.h"
#include "tensorflow_lite_support/examples/task/vision/desktop/utils/image_utils_c.h""


struct ImageClassifierOptions* BuildImageClassifierOptions(const char* model_path) {

  ImageClassifierOptions *options = ImageClassifierOptionsCreate();
  // ImageClassifierOptionsAddClassNameBlackList(options, "/m/01bwbt");
  // ImageClassifierOptionsAddClassNameBlackList(options, "/m/0bwm6m");
  ImageClassifierOptionsSetScoreThreshold(options, 0);
  ImageClassifierOptionsSetMaxResults(options, 3);
  ImageClassifierOptionsSetModelFilePath(options, model_path);

  return options;
}
 
void DisplayClassificationResults(struct ClassificationResult *classification_result) {

  for (int head = 0; head < classification_result->size; ++head) {
      struct Classifications test_classification = classification_result->classifications[head];
      for (int rank = 0; rank < test_classification.size; ++rank) {
          printf("  Rank #%d:\n", rank);
          printf("   index       : %d\n",
                                        test_classification.classes[rank].index);
          printf("   score       : %.5f\n",
                                        test_classification.classes[rank].score);

          printf("   display name: %s\n",
                                  test_classification.classes[rank].display_name);
          printf("   class name  : %s\n",
                                          test_classification.classes[rank].class_name);
    }
  }
}

void Classify(const char* model_path, const char* image_path) {
  struct ImageClassifierOptions *options = BuildImageClassifierOptions(model_path);
  ImageClassifier *image_classifier = ImageClassifierFromOptions(options);
  if (image_classifier == NULL) {
    printf("An error occured while instantiating the Image Classifier\n");
    return;
  }
  
  struct ImageData image_data = DecodeImageFromFile(image_path);
  
  struct FrameBuffer frame_buffer = {.dimension.width = image_data.width, 
                                     .dimension.height = image_data.height, 
                                     .plane.buffer = image_data.pixel_data, 
                                     .plane.stride.row_stride_bytes = image_data.width  * image_data.channels, 
                                     .plane.stride.pixel_stride_bytes = image_data.channels, 
                                     .format = kRGB};
  
  struct ClassificationResult *classification_result = ImageClassifierClassify(image_classifier, &frame_buffer);
  ImageClassifierDelete(image_classifier);

  if (classification_result == NULL) {
    printf("An error occured while classifying the image\n");
    return;
  }
  DisplayClassificationResults(classification_result);
  
  ImageClassifierClassificationResultDelete(classification_result);
  ImageDataFree(&image_data);
}

int main(int argc, char** argv) {
  const char *model_path = "/tmp/aiy_vision_classifier_birds_V1_3.tflite";
  const char *image_path = "/Users/priankakariat/Documents/Projects/TensorFlow/tflite-support/tensorflow_lite_support/examples/task/vision/desktop/g3doc/sparrow.jpg";
  Classify(model_path, image_path);
  return 0;

}

File: tensorflow_lite_support/examples/task/vision/desktop/python/image_classifier_demo.py
# Copyright 2021 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
"""Python demo tool for Image Classification."""

import json

from absl import app
from absl import flags
from google.protobuf import json_format

from tensorflow_lite_support.python.task.core import task_options
from tensorflow_lite_support.python.task.processor.proto import classification_options_pb2
from tensorflow_lite_support.python.task.vision import image_classifier
from tensorflow_lite_support.python.task.vision.core import tensor_image

FLAGS = flags.FLAGS

flags.DEFINE_string('model_path', None,
                    'Absolute path to the ".tflite" image classifier model.')
flags.DEFINE_string(
    'image_path', None,
    'Absolute path to the image to classify. The image must be RGB or '
    'RGBA (grayscale is not supported). The image EXIF orientation '
    'flag, if any, is NOT taken into account.')
flags.DEFINE_integer('max_results', 5,
                     'Maximum number of classification results to display.')
flags.DEFINE_float(
    'score_threshold', 0,
    'Classification results with a confidence score below this value are '
    'rejected. If >= 0, overrides the score threshold(s) provided in the '
    'TFLite Model Metadata. Ignored otherwise.')
flags.DEFINE_string(
    'class_name_allowlist', '',
    'Comma-separated list of class names that acts as a whitelist. If '
    'non-empty, classification results whose "class_name" is not in this list '
    'are filtered out. Mutually exclusive with "class_name_denylist".')
flags.DEFINE_string(
    'class_name_denylist', '',
    'Comma-separated list of class names that acts as a blacklist. If '
    'non-empty, classification results whose "class_name" is in this list '
    'are filtered out. Mutually exclusive with "class_name_allowlist".')
flags.DEFINE_bool(
    'use_coral', False,
    'If true, inference will be delegated to a connected Coral Edge TPU '
    'device.')


def build_options():
  base_options = task_options.BaseOptions(
    model_file=task_options.ExternalFile(file_name=FLAGS.model_path),
    use_coral=FLAGS.use_coral)
  classification_options = classification_options_pb2.ClassificationOptions(
    max_results=FLAGS.max_results, score_threshold=FLAGS.score_threshold,
    class_name_allowlist=FLAGS.class_name_allowlist,
    class_name_denylist=FLAGS.class_name_denylist, use_coral=FLAGS.use_coral)
  return image_classifier.ImageClassifierOptions(
    base_options=base_options, classification_options=classification_options)


def main(_) -> None:
  # Creates classifier.
  options = build_options()
  classifier = image_classifier.ImageClassifier.create_from_options(options)

  # Loads image.
  image = tensor_image.TensorImage.from_file(FLAGS.image_path)

  # Run classification.
  result = classifier.classify(image)

  # Gets results.
  image_result_dict = json.loads(json_format.MessageToJson(result))
  print(image_result_dict)


if __name__ == "__main__":
  flags.mark_flag_as_required('model_path')
  flags.mark_flag_as_required('image_path')
  app.run(main)
